msgspec/__init__.pyi

```diff
@@ -19,10 +19,15 @@ from . import inspect, json, msgpack, structs, toml, yaml
 
 T = TypeVar("T")
 
-class _Unset:
+class UnsetType:
     pass
 
-UNSET = _Unset()
+UNSET = UnsetType()
```

Comment:
If `UNSET` is implemented as an object of an empty `UnsetType` class, static type checkers will unfortunately have a hard time doing type narrowing on it. I'd consider instead making `UNSET` the single value of an enum, so `UnsetType` can be `Literal[UNSET]` and type narrowing will work.

[Example of the problem and how the enum trick fixes it.](https://gist.github.com/SyntaxColoring/d940e6bc5cd37e499779ae8d527c9da1)

This is [GVR's suggested workaround](https://github.com/python/typing/issues/689#issuecomment-561425237). Hopefully someday [PEP 661](https://peps.python.org/pep-0661/) will give us standard sentinel objects, and we'll be able to get this to work without enum trickery. :)